SimPEG/electromagnetics/base_1d_stitched.py

```diff
@@ -0,0 +1,544 @@
+        return self._IJLayers
+
+    @property
+    def IJHeight(self):
+        if getattr(self, "_IJHeight", None) is None:
+            # Ordering: first z then x
+            self._IJHeight = self.set_ij_n_layer(n_layer=1)
+        return self._IJHeight
+
+    # ------------- For physics ------------- #
+
+    def get_uniq_soundings(self):
+        self._sounding_types_uniq, self._ind_sounding_uniq = np.unique(
+            self.survey._sounding_types, return_index=True
+        )
+
+    def input_args(self, i_sounding, output_type="forward"):
+        output = (
+            self.survey.get_sources_by_sounding_number(i_sounding),
+            self.topo[i_sounding, :],
+            self.Thicknesses[i_sounding, :],
+            self.Sigma[i_sounding, :],
+            self.Eta[i_sounding, :],
+            self.Tau[i_sounding, :],
+            self.C[i_sounding, :],
+            self.Chi[i_sounding, :],
+            self.dChi[i_sounding, :],
+            self.Tau1[i_sounding, :],
+            self.Tau2[i_sounding, :],
+            self.H[i_sounding],
+            output_type,
+            # False,
+            # self._coefficients[i_sounding],
+        )
+        return output
+
+    # This is the most expensive process, but required once
+    # May need to find unique
+    # def input_args_for_coeff(self, i_sounding):
+    #     output = (
+    #         self.survey.get_sources_by_sounding_number(i_sounding),
+    #         self.topo[i_sounding, :],
+    #         self.Thicknesses[i_sounding,:],
+    #         self.Sigma[i_sounding, :],
+    #         self.Eta[i_sounding, :],
+    #         self.Tau[i_sounding, :],
+    #         self.C[i_sounding, :],
+    #         self.Chi[i_sounding, :],
+    #         self.dChi[i_sounding, :],
+    #         self.Tau1[i_sounding, :],
+    #         self.Tau2[i_sounding, :],
+    #         self.H[i_sounding],
+    #         'forward',
+    #         True,
+    #         [],
+    #     )
+    #     return output
+
+    def fields(self, m):
+        if self.verbose:
+            print("Compute fields")
+
+        return self.forward(m)
+
+    def dpred(self, m, f=None):
+        """
+        Return predicted data.
+        Predicted data, (`_pred`) are computed when
+        self.fields is called.
+        """
+        if f is None:
+            f = self.fields(m)
+
+        return f
+
+    @property
+    def sounding_number(self):
+        self._sounding_number = [
+            key for key in self.survey.source_location_by_sounding_dict.keys()
+        ]
+        return self._sounding_number
+
+    @property
+    def n_chunk(self):
+        self._n_chunk = len(self.sounding_number_chunks)
+        return self._n_chunk
+
+    @property
+    def source_locations_for_sounding(self):
+        if getattr(self, "_source_locations_for_sounding", None) is None:
+            self._source_locations_for_sounding = np.vstack(
+                [
+                    self.survey._source_location_by_sounding_dict[ii][0]
```

Comment:
See comment below about changing the dictionary, so you can do directly `np.vstack(self.surveys._source_location_by_sounding.value)`